core/include/gnuradio-4.0/Graph.hpp

```diff
@@ -780,29 +812,87 @@ struct Graph : public gr::Block<Graph> {
-    void forEachBlockMutable(F&& f) {
-        std::ranges::for_each(_blocks, [f](auto& block_ptr) { std::invoke(f, *block_ptr.get()); });
-    }
+static_assert(BlockLike<Graph>);
 
-    template<std::invocable<Edge&> F>
-    void forEachEdgeMutable(F&& f) {
-        std::ranges::for_each(_edges, f);
-    }
+/*******************************************************************************************************/
+/**************************** Graph helper functions ***************************************************/
+/*******************************************************************************************************/
 
-    template<std::invocable<const BlockModel&> F>
-    void forEachBlock(F&& f) const {
-        std::ranges::for_each(_blocks, [f](auto& block_ptr) { std::invoke(f, std::as_const(*block_ptr.get())); });
-    }
+namespace graph {
 
-    template<std::invocable<const Edge&> F>
-    void forEachEdge(F&& f) const {
-        std::ranges::for_each(_edges, f);
-    }
-};
+namespace detail {
+template<block::Category traverseCategory, typename Fn>
+void traverseSubgraphs(GraphLike auto const& root, Fn&& visitGraph) {
+    using enum block::Category;
 
-static_assert(BlockLike<Graph>);
+    auto recurse = [&visitGraph](GraphLike auto const& graph, auto& self) -> void {
+        visitGraph(graph);
+
+        for (const auto& block : graph.blocks()) {
+            const auto cat = block->blockCategory();
+            if constexpr (traverseCategory == All) {
+                if (cat == TransparentBlockGroup || cat == ScheduledBlockGroup) { // block is a sub-graph
+                    self(*block, self);
+                }
+            } else if (cat == traverseCategory) {
+                self(*block, self);
+            }
+        }
+    };
+    recurse(root, recurse);
+}
+
+} // namespace detail
+
+template<block::Category traverseCategory, typename Fn>
+requires(std::invocable<Fn, const std::shared_ptr<BlockModel>> || std::invocable<Fn, std::shared_ptr<BlockModel>>)
+void forEachBlock(GraphLike auto const& root, Fn&& function, block::Category filter = block::Category::All) {
+    using enum block::Category;
+    using BlockType = std::conditional_t<std::invocable<Fn, const std::shared_ptr<BlockModel>>, const std::shared_ptr<BlockModel>&, std::shared_ptr<BlockModel>&>;
+
+    detail::traverseSubgraphs<traverseCategory>(root, [&](GraphLike auto const& graph) {
+        for (BlockType block : graph.blocks()) {
+            const block::Category cat = block->blockCategory();
+            if (filter == All || cat == filter) {
+                function(block);
+            }
+        }
+    });
+}
+
+template<block::Category traverseCategory, typename Fn>
+requires(std::invocable<Fn, const Edge&> || std::invocable<Fn, Edge&>)
+void forEachEdge(GraphLike auto const& root, Fn&& function, Edge::EdgeState filter) {
+    using enum Edge::EdgeState;
+    using EdgeType = std::conditional_t<std::invocable<Fn, const Edge&>, const Edge&, Edge&>;
```

Comment:
This will not detect properly const vs non-const for generic lambdas
https://godbolt.org/z/v75d5cxv7

Reply:
Thanks for the feedback. I’ve dropped the `requires(...)` clause and type conditional, as it doesn’t reliably improve diagnostics, especially with generic lambdas where `std::invocable<Fn, const T&>` may yield false positives. The compiler provides a meaningful error at the point of invocation anyway.